pkg/models/devops/devops.go

```diff
@@ -1009,3 +1014,118 @@ func parseBody(body io.Reader) (newReqBody io.ReadCloser) {
+	}
+	var allParams []devopsv1alpha3.ParameterDefinition
+	allParams = append(allParams, piplineObj.Spec.Pipeline.Parameters...)
+	for _, param := range piplineObj.Spec.Pipeline.ParametersFrom {
+		if param.Kind == "ConfigMap" {
+			if param.Mode == devopsv1alpha3.PARAM_REF_MODE_CONFIG {
+				params, err := buildParamFromConfigMapData(d.k8sclient, param.Name, param.ValuesKey)
+				if err != nil {
+					klog.Errorf("buildParamFromConfigMapData error:[%v]", err)
+					continue
+				}
+				allParams = append(allParams, params...)
+			} else if param.Mode == devopsv1alpha3.PARAM_REF_MODE_RESTFUL {
+				params, err := buildParamFromRESTfull(d.k8sclient, param.Name, query)
+				if err != nil {
+					klog.Errorf("buildParamFromRESTfull error:[%v]", err)
+					continue
+				}
+				allParams = append(allParams, params...)
+			}
+		}
+	}
+	// remove items with duplicated name
+	ret := mergeParameters(allParams)
+
+	return ret, nil
+}
+
+func mergeParameters(params []devopsv1alpha3.ParameterDefinition) []devopsv1alpha3.ParameterDefinition {
+	var ret []devopsv1alpha3.ParameterDefinition
+	tmpMap := make(map[string]byte)
+	for i := 0; i < len(params); i++ {
+		idx := len(params) - 1 - i
+		cfgName := params[idx].Name
+		if _, ok := tmpMap[cfgName]; !ok {
+			ret = append(ret, params[idx])
+			tmpMap[cfgName] = 1
+		}
+	}
+	return ret
+}
+
+func checkParametersCMName(fullName string) (namespace string, name string, err error) {
+	nameInfo := strings.SplitN(name, "/", 2)
+	if len(nameInfo) != 2 {
+		return "", "", fmt.Errorf("invalid name [%s]", name)
+	}
+	return nameInfo[0], nameInfo[1], nil
+}
+
+func buildParamFromConfigMapData(k8sclient kubernetes.Interface, name, valuesKey string) ([]devopsv1alpha3.ParameterDefinition, error) {
+	var ret []devopsv1alpha3.ParameterDefinition
+	cmNS, cmName, err := checkParametersCMName(name)
+	if err != nil {
+		return ret, err
+	}
+	ctx := context.Background()
+	cm, err := k8sclient.CoreV1().ConfigMaps(cmNS).Get(ctx, cmName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if paramData, ok := cm.Data[valuesKey]; ok {
+		err := yaml.Unmarshal([]byte(paramData), &ret)
+		return ret, err
+	}
+	return nil, fmt.Errorf("data not found in configmap with key [%s]", valuesKey)
+
+}
+
+func buildParamFromRESTfull(k8sclient kubernetes.Interface, name string, query url.Values) ([]devopsv1alpha3.ParameterDefinition, error) {
+	var ret []devopsv1alpha3.ParameterDefinition
+	cmNS, cmName, err := checkParametersCMName(name)
+	if err != nil {
+		return ret, err
+	}
+	ctx := context.Background()
+	cm, err := k8sclient.CoreV1().ConfigMaps(cmNS).Get(ctx, cmName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	if u, ok := cm.Data["url"]; ok {
+		_, err := url.Parse(u)
+		if err != nil || len(u) == 0 {
+			return ret, fmt.Errorf("invalid url [%s] in configmap[%s]", u, name)
+		}
+		client := resty.New()
+		// request from url
+		resp, err := client.R().SetQueryParamsFromValues(query).Get(u)
```

Comment:
Currently, it only sends very simple HTTP GET requests. I prefer to use the built-in method in Golang. We can import an HTTP library when we need it. What do you think?

Reply:
I prefer to expose more options to make use of this package

Reply:
> I prefer to expose more options to make use of this package

Yes, of course. But it does not have more options now. Right? For now, a simple HTTP client is good enough. I didn't see any reason to import an advanced library now.

Reply:
ok